File: toggle_led_cpp/src/battery_state.cpp
#include "rclcpp/rclcpp.hpp"
#include "led_interface/srv/led_state.hpp"

class BatteryState: public rclcpp::Node{

    private:

        std::thread thread_;
        rclcpp::TimerBase::SharedPtr timer_;
        std::string battery_state;
        float last_time_state_changed;

        double get_current_time(){

            auto currentTimePoint = std::chrono::system_clock::now();

            // Convertendo o tempo para um formato de tempo desde o epoch
            auto duration = currentTimePoint.time_since_epoch();

            // Obtendo o tempo em segundos e nanossegundos
            auto secs = std::chrono::duration_cast<std::chrono::seconds>(duration);
            auto nsecs = std::chrono::duration_cast<std::chrono::nanoseconds>(duration - secs);

            std::cout << secs.count()<< std::endl;
            return secs.count();

        }

        void call_server(int led_number, std::string state){
            
            auto client = this->create_client<led_interface::srv::LedState>("set_led");
            auto request = std::make_shared<led_interface::srv::LedState::Request>();

            while (!client->wait_for_service(std::chrono::seconds(1)))
                RCLCPP_WARN(this->get_logger(), "Waiting for service");

            request->led_number = led_number;
            request->state = state;

            auto future = client->async_send_request(request);

            try{
                
                auto response = future.get();
                RCLCPP_INFO(this->get_logger(), "Success: %d", response->success);

            }

            catch (std::exception &ex){
                RCLCPP_ERROR(this->get_logger(), "Exception occurred");
            }
        }


        void check_state(){
            float time_now = get_current_time();
            int led_number = 3;
            if(battery_state == "full"){
                if(time_now - last_time_state_changed > 4){
                    battery_state = "empty";
                    this->call_server(led_number, battery_state);
                    RCLCPP_WARN(this->get_logger(), "Battery is empty... Charging");
                    last_time_state_changed = time_now;
                }
            }

            else{
                    battery_state = "full";
                    this->call_server(led_number, battery_state);
                    RCLCPP_INFO(this->get_logger(), "Battery is full again!");
                    last_time_state_changed = time_now;
            }
        }

    public:

        BatteryState(): Node("battery_node"){

            thread_ = std::thread(std::bind(&BatteryState::call_server, this, 3, "empty"));
            timer_ = this->create_wall_timer(std::chrono::seconds(1), std::bind(&BatteryState::check_state, this));
            battery_state = "full";
            last_time_state_changed = this->get_current_time();
        }
};

int main(int argc, char ** argv){

    rclcpp::init(argc, argv);
    auto node = std::make_shared<BatteryState>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;

}

File: python_packages/toggle_led_python/toggle_led_python/battery_state.py
import rclpy
from rclpy.node import Node
from led_interface.srv import LedState
from functools import partial

class BatteryState(Node):
    def __init__(self):
        super().__init__("battery_state")
        self.battery_state = "full"
        self.last_time_state_changed = self.get_current_time()
        self.timer_ = self.create_timer(0.1, self.check_battery_state)
        self.get_logger().info("Battery state has been initialized")

    def get_current_time(self):
        secs, nsecs = self.get_clock().now().seconds_nanoseconds()

        return secs + nsecs/1000000000.0
    
    def check_battery_state(self):
        time_now = self.get_current_time()
        if self.battery_state == "full":
            if time_now - self.last_time_state_changed > 4:
                self.battery_state = "empty"
                self.call_server(1, self.battery_state)
                self.last_time_state_changed = time_now
                self.get_logger().info("Battery is empty: charging...")

        else:
            if time_now - self.last_time_state_changed > 6:
                self.battery_state = "full"
                self.call_server(1, self.battery_state)
                self.last_time_state_changed = time_now
                self.get_logger().info("Battery is now full again")
             

    def call_server(self, led_number: int, state: str):
       
        """
        led_number: led from 1 to 3
        state: string that can be "full" or "empty"
        """

        self.client_ = self.create_client(LedState, "set_led")
        request = LedState.Request()
        request.led_number = led_number
        request.state = state

        while not self.client_.wait_for_service(1):
            self.get_logger().warn("Waiting for service")

        future = self.client_.call_async(request)
        future.add_done_callback(partial(self.callback_call))

    def callback_call(self, future):
        try:
            response = future.result()
            self.get_logger().info(f"Success: {response.success}")

        except Exception as ex:
            self.get_logger().error("Something wrong is not correct %r" %(ex,))


def main(args=None):
    rclpy.init(args=args)
    node = BatteryState()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()